src/main/scala/cognite/spark/v1/FlexibleDataModelCorePropertySyncRelation.scala

```diff
@@ -110,54 +110,111 @@ private[spark] class FlexibleDataModelCorePropertySyncRelation(
                 p != "startNode" && p != "endNode" && p != "space" && p != "externalId" && p != "type")
         ))
       .toSeq
+    val cursors = if (cursor.nonEmpty) Some(Map("sync" -> cursor)) else None
+    val select = Map("sync" -> SelectExpression(sources = sourceReference))
+    val (futureItemsCursor, executeBackFill) =
+      generateFutureItemsCursor(cursors, instanceType, select).unsafeRunSync()
+
+    Seq(
+      syncOut(
+        futureItemsCursor,
+        cursors = if (executeBackFill) None else cursors,
+        `with` = Map("sync" -> tableExpression),
+        select,
+        selectedColumns
+      ))
+  }
 
-    val instanceSyncRequest = InstanceSyncRequest(
-      `with` = Map("sync" -> tableExpression),
-      cursors = if (cursor.nonEmpty) Some(Map("sync" -> cursor)) else None,
-      select = Map("sync" -> SelectExpression(sources = sourceReference)),
-      includeTyping = Some(true)
-    )
+  /**
+    * Generate future items cursor and validate cursor expiration.
+    * If the cursor has expired, generate a new cursor and do a full back fill.
+    */
+  private def generateFutureItemsCursor(
+      cursors: Option[Map[String, String]],
+      instanceType: InstanceType,
+      select: Map[String, SelectExpression]): IO[(Option[String], Boolean)] =
+    fetchData(
+      isBackFill = false,
+      cursors = cursors,
+      `with` = Map(
+        "sync" -> generateTableExpression(instanceType, FilterDefinition.Not(MatchAll(JsonObject())))),
+      select = select)
+      .map { sr =>
+        (sr.nextCursor, cursors.isEmpty)
+      }
+      .redeemWith(
+        {
+          case e: CdpApiException if e.code == 400 && e.message.startsWith("Cursor has expired") =>
+            logger.warn("Cursor has expired, doing a full backfill")
+            generateFutureItemsCursor(None, instanceType, select)
```

Comment:
With expired cursors it will call into fetchData using sync endpoint, is it expected? Then it will return `(sr.nextCursor, true)` which it seems should mean that called needs to use `query` endpoint to do a backfill starting from `sr.nextCursor`? But is cursor returned from /sync a valid cursor for /query?

Reply:
The backfill will start from empty cursor, and then continue cursoring on cursor by query. However the cursor string we return from the client will be future items cursor from sync, which will be used on next sync session..